tests/vfolders/path_matcher.py

```diff
@@ -0,0 +1,194 @@
+    assert (
+        sorted(vfolder.path_matcher.languages)
+        == [language0.pk, language1.pk])
+    vfolder.all_languages = True
+    vfolder.save()
+    assert vfolder.path_matcher.languages is None
+    assert list(vfolder.path_matcher.projects) == []
+    # no projects set so no stores
+    assert list(vfolder.path_matcher.existing_stores) == []
+    vfolder.all_languages = False
+    vfolder.save()
+    # didnt forget lang preferences
+    vfolder.all_languages = False
+    vfolder.save()
+    assert (
+        sorted(vfolder.path_matcher.languages.values_list("code", flat=True))
+        == ["language0", "language1"])
+
+
+@pytest.mark.pootle_vfolders
+@pytest.mark.django_db
+def test_vfolder_path_matcher_projects(project0, project1):
+    vfolder = VirtualFolder.objects.create(
+        name="avfolder",
+        filter_rules="FOO,BAR")
+    vfolder.projects.add(project0)
+    vfolder.projects.add(project1)
+    # no languages set so no stores
+    assert list(vfolder.path_matcher.existing_stores) == []
+    assert (
+        sorted(vfolder.path_matcher.projects.values_list("code", flat=True))
+        == ["project0", "project1"])
+    vfolder.all_projects = True
+    vfolder.save()
+    assert vfolder.path_matcher.projects is None
+    assert list(vfolder.path_matcher.languages) == []
+    # no languages set so no stores
+    assert list(vfolder.path_matcher.existing_stores) == []
+    vfolder.all_projects = False
+    vfolder.save()
+    # didnt forget lang preferences
+    vfolder.all_projects = False
+    vfolder.save()
+    assert (
+        sorted(vfolder.path_matcher.projects.values_list("code", flat=True))
+        == ["project0", "project1"])
+
+
+@pytest.mark.pootle_vfolders
+@pytest.mark.django_db
+def test_vfolder_path_matcher_all_proj_lang():
+    vfolder = VirtualFolder.objects.create(
+        name="avfolder",
+        all_projects=True,
+        all_languages=True,
+        filter_rules="*")
+    return
+    assert (
+        list(vfolder.path_matcher.existing_stores.order_by("pk"))
+        == list(Store.objects.order_by("pk")))
+    vfolder.all_languages = False
+    vfolder.save()
+    assert (
+        list(vfolder.path_matcher.existing_stores.order_by("pk"))
+        == [])
+    vfolder.all_languages = True
+    vfolder.all_languages = False
+    vfolder.save()
+    assert (
+        list(vfolder.path_matcher.existing_stores.order_by("pk"))
+        == [])
+
+
+@pytest.mark.pootle_vfolders
+@pytest.mark.django_db
+def test_vfolder_path_matcher_some_proj_lang(tp0, language1):
+    vfolder = VirtualFolder.objects.create(
+        name="avfolder",
+        filter_rules="*")
+    vfolder.languages.add(tp0.language)
+    vfolder.projects.add(tp0.project)
+    # m2m handler?
+    vfolder.save()
+    assert (
+        list(vfolder.path_matcher.existing_stores.order_by("pk"))
+        == list(tp0.stores.order_by("pk")))
+    vfolder.languages.add(language1)
+    vfolder.save()
+    assert (
```

Comment:
Also add testing for single language and all projects, and for all projects and single language.